annotations/annotations_api.go

```diff
@@ -2,37 +2,145 @@ package annotations
+	UPPNotFoundMsg = "UPP responded with not found"
+	// UPPServiceUnavailableMsg is error message used when UPP annotations endpoint returns http error code
+	// different from bad request and not found
+	UPPServiceUnavailableMsg = "Service unavailable"
+
+	pacAnnotationLifecycle       = "pac"
+	v1AnnotationLifecycle        = "v1"
+	nextVideoAnnotationLifecycle = "next-video"
+)
+
+// UPPError encapsulate error information for errors originating from calls to UPP annotations endpoint.
+type UPPError struct {
+	msg     string
+	status  int
+	uppBody []byte
+}
 
-const syntheticContentUUID = "4f2f97ea-b8ec-11e4-b8e6-00144feab7de"
+// Error returns the error message.
+func (ue UPPError) Error() string {
+	return ue.msg
+}
 
-type UPPAnnotationsAPI interface {
-	Get(ctx context.Context, contentUUID string) (*http.Response, error)
-	Endpoint() string
-	GTG() error
+// Status returns the http status code returned by the call to the UPP annotations endpoint.
+func (ue UPPError) Status() int {
+	return ue.status
 }
 
-type annotationsAPI struct {
+// UPPBody returns the http response body returned by the call to the UPP annotations endpoint.
+func (ue UPPError) UPPBody() []byte {
+	return ue.uppBody
+}
+
+// UPPAnnotationsAPI retrieves published annotations from UPP.
+type UPPAnnotationsAPI struct {
 	endpointTemplate string
 	apiKey           string
 	httpClient       *http.Client
 }
 
-func NewUPPAnnotationsAPI(client *http.Client, endpoint string, apiKey string) UPPAnnotationsAPI {
-	return &annotationsAPI{endpointTemplate: endpoint, apiKey: apiKey, httpClient: client}
+// NewUPPAnnotationsAPI initializes UPPAnnotationsAPI by given http client,
+// the url of the UPP public endpoint for getting published annotations and UPP API key.
+func NewUPPAnnotationsAPI(client *http.Client, endpoint string, apiKey string) *UPPAnnotationsAPI {
+	return &UPPAnnotationsAPI{endpointTemplate: endpoint, apiKey: apiKey, httpClient: client}
+}
+
+// GetAll retrieves the list of published annotations for given contentUUID.
+// The returned list contains the annotations returned by UPP without filtering.
+func (api *UPPAnnotationsAPI) GetAll(ctx context.Context, contentUUID string) ([]Annotation, error) {
+	return api.getAnnotations(ctx, contentUUID)
+}
+
+// GetAllButV2 retrieves the list of published annotations for given contentUUID but filtering v2 annotations.
+func (api *UPPAnnotationsAPI) GetAllButV2(ctx context.Context, contentUUID string) ([]Annotation, error) {
+	return api.getAnnotations(ctx, contentUUID, pacAnnotationLifecycle, v1AnnotationLifecycle, nextVideoAnnotationLifecycle)
+}
+
+func (api *UPPAnnotationsAPI) getAnnotations(ctx context.Context, contentUUID string, lifecycles ...string) ([]Annotation, error) {
+	uppResponse, err := api.getUPPAnnotationsResponse(ctx, contentUUID, lifecycles...)
+	if err != nil {
+		return nil, err
+	}
+
+	defer uppResponse.Body.Close()
+	respBody, err := ioutil.ReadAll(uppResponse.Body)
```

Comment:
I have to send the response body in case of 400 and 404 and that's why the code for reading the response is before checking the status code. I have to send it because the service was written that way before the refactoring and I didn't want to introduce any changes in the expected behaviour.